### ISSUE TITLE
Problem with running simple toy problem script

### ISSUE THREAD

**Issue by amitfishy** [jeanfeydy/geomloss#31]
Hello,

I'm trying to run a simple toy problem, where I have a target Multivariate Normal and a tunable mean, std Multivariate to try to match the target. The loss would be the sinkhorn loss between the samples from the target and the samples from the tunable model and the mean and std should be optimized by something like the ADAM optimizer. But it does not seem like the gradients are being back-propagated to the mean/std parameters to modify their values (when I check values using torch.autograd the gradients come out as None). I'm not sure what I'm doing wrong. Any help would be appreciated! Thanks!

```
import torch
import torch.nn as nn
import torch.optim as optim

from torch.distributions.multivariate_normal import MultivariateNormal

from geomloss import SamplesLoss

DIM_SIZE = 3
NUM_SAMPLES = 64
LR = 0.01
NUM_STEPS = 2000

target_mean = torch.ones(DIM_SIZE)
target_std = torch.ones(DIM_SIZE)*0.25
target_m = MultivariateNormal(target_mean, torch.diag(target_std))

mean = nn.Parameter(torch.zeros(DIM_SIZE, requires_grad = True))
std = nn.Parameter(torch.ones(DIM_SIZE, requires_grad = True))
optimizer = optim.Adam([mean, std], lr=LR)

loss = SamplesLoss(loss="sinkhorn", p=2, blur=.05)

for i in range(NUM_STEPS):
	m = MultivariateNormal(mean, torch.diag(std))

	target_samples = target_m.sample(sample_shape=torch.Size([NUM_SAMPLES]))
	param_samples = m.sample(sample_shape=torch.Size([NUM_SAMPLES])).requires_grad_()

	sinkhorn_loss = loss(param_samples, target_samples)
	# sinkhorn_loss = sinkhorn_loss.sum()
	optimizer.zero_grad()
	sinkhorn_loss.backward()
	optimizer.step()

	# g_mean, g_std = torch.autograd.grad(sinkhorn_loss, [mean, std])
	# print (g_mean, g_std)
	
	print('Error on mean: {} std: {} Sinkhorn Loss: {}'.format(torch.mean(torch.abs(target_mean-mean.cpu())), torch.mean(torch.abs(target_std-std.cpu())), sinkhorn_loss))
```

**Comment by wzm2256** [comment]
Did you try rsample instead of sample? i.e. m.rsample instead of m.sample. sample does not have gradient.

**Comment by amitfishy** [comment]
Thanks for the help! Yes that seems to work.

However, if I replace the sinkhorn loss here by a log likelihood loss of the `param_samples`, the gradient is backpropagated to the parameters `mean` and `std` simply by using `sample`.

Consider the very similar script:
```
import torch
import torch.nn as nn
import torch.optim as optim

from torch.distributions.multivariate_normal import MultivariateNormal


def get_kl_loss(target_samples, param_samples, param_dist):
	kl_loss = - torch.mean(param_dist.log_prob(target_samples))
	return kl_loss


DIM_SIZE = 16
NUM_SAMPLES = 32
LR = 0.001
NUM_STEPS = 2000

target_mean = torch.ones(DIM_SIZE)
target_std = torch.ones(DIM_SIZE)*0.25
target_m = MultivariateNormal(target_mean, torch.diag(target_std))

mean = nn.Parameter(torch.zeros(DIM_SIZE))
std = nn.Parameter(torch.ones(DIM_SIZE))
optimizer = optim.Adam([mean, std], lr=LR)



for i in range(NUM_STEPS):
	m = MultivariateNormal(mean, torch.diag(std))

	target_samples = target_m.sample(sample_shape=torch.Size([NUM_SAMPLES]))
	param_samples = m.sample(sample_shape=torch.Size([NUM_SAMPLES]))

	kl_loss = get_kl_loss(target_samples, param_samples, m)

	optimizer.zero_grad()
	kl_loss.backward()
	optimizer.step()

	print('Error on mean: {} std: {} kl: {}'.format(torch.mean(torch.abs(target_mean-mean)), torch.mean(torch.abs(target_std-std)), kl_loss))
```

I'm just trying to figure out why I didn't have to use `rsample` in this case for it to work, because it seems like the gradients are being back-propagated in this case even though I just used `sample`.
